Express-rate-limit Integration

rpigneri-vol opened the pull request:
Since this is my first check-in, I set a pretty low bar for integrating express-rate-limit.  Specifically, I'm limiting every single resource to 100 requests per 15 minutes.  At that rate, I have little concern of creating false alerts because we are limiting our synthetics.

That being said, I will need to update this (nearly immediately) to handle multi-server scaling sets.

rpigneri-vol committed:
Since this is my first check-in, I set a pretty low bar for integrating express-rate-limit.  Specifically, I'm limiting every single resource to 100 requests per 15 minutes.  This should easily clear our synthetics.

That being said, I will need to update this (nearly immediately) to handle multi-server scaling sets.

johndeyrup commented on app.js:
```diff
@@ -88,6 +88,15 @@ if (cluster.isMaster) {
 
+        // Set rate limiting for all features as a first cut.
+        // This will most probably need some refinement for specific endpoints and multi-server scaling.
+        const rateLimit = require('express-rate-limit');
```
The docs recommend setting `app.set('trust proxy', 1);` if you are using elastic bean stalk https://www.npmjs.com/package/express-rate-limit

rpigneri-vol replied:
Hi, @johndeyrup ,

The comments are specific: use that line for ELB, but only if you are using a reverse proxy.  There is more detail at https://expressjs.com/en/guide/behind-proxies.html on how to refine express-rate-limit based upon how the proxy operates.  How can I find out how the reverse proxy is configured so I'm not working against it?

Thank you,

Rocco

johndeyrup replied:
I can't find any docs about whether ELB uses a reverse proxy by default. Did you come across anything when you set it up @mzagaja ?

mzagaja replied:
@johndeyrup We had a conversation about this and think the nginx server in the EC2 instances constitute a "reverse proxy" for this purpose.

rpigneri-vol replied:
I have updated the check-in per our conversation.  I chose the "trust proxy" Boolean setting rather than the per-IP setting because the per-IP setting seems like it might accidentally group clients together.   The Boolean setting seemed more reasonable.

rpigneri-vol committed:
Configured express-rate-limit to "trust the proxy".

This seemed a more reasonable setup than providing a list of IP addresses as the IP addresses might end up grouping the clients rather than identifying them accurately.  This is because the right-most address after the proxy IP's is used as the client IP in the X-Forwarded-For header.

rpigneri-vol committed:
Checking in the test harness so others can use it.
